### DISCUSSION TITLE
Exact algorithm of inverse problem of PINNs

### DISCUSSION THREAD

**Discussion by shanmukhi1998** [lululxvi/deepxde#1272] [General]
Hi Dr Lu

Thank you for such a wonderful library. I have been using it for solving inverse problems. However, while solving the problems, I did encounter a few questions. It would be of great help if you could clarify them.
1. What is the exact algorithm that happens when solving an inverse problem? I noticed that in my case, Neural network loss has converged to a certain value but PDE loss continues to decrease. Earlier, I thought that Inverse PINNS functions as a regular deep learning model except that the loss function is now a sum of PDE loss and NN loss. If PDE loss is changing, the overall loss is also changing and hence, the NN weights must change. This should lead to changing NN loss too, right? But it doesn't happen. Am I missing something here? It would be helpful if you could explain the exact algorithm of inverse PINNS problem or maybe point me to some resource that explains the same.
2. Also, I have currently put my learning rate as 1e-04 in my problem which is using adam optimizer. After 6000 iterations or so, I see jumps in PDE loss. Do you think, further reducing the learning rate will help?

Kindly clarify these. Thank you.

**Comment by pescap** [comment]
Hi, could you show us a convergence print of how both losses behave?

**Reply by shanmukhi1998** [reply]
Hi

Please find it below. I would like to clarify certain things:
I have put anchors=X (which is my entire dataset) (num_domain = 0 and have not explicitly specified num_test), my architecture is: [3]-[18]*16-[1] (3 inputs and 1 output), learning rate: 1e-04. 
I appreciate your help. Thanks.

[loss.txt](https://github.com/lululxvi/deepxde/files/11347977/loss.txt)

**Reply by pescap** [reply]
`num_domain = 0` ?

**Reply by shanmukhi1998** [reply]
Yeah. Because I wanted to evaluate PDE residual only at the points that I specify

**Reply by pescap** [reply]
The loss terms are highly unbalanced. Did you try to try to change the `loss_weights`?

Example: https://github.com/lululxvi/deepxde/blob/987f52671a4774f13ec3035d7e49426f99e4d08d/examples/pinn_forward/Helmholtz_Sound_hard_ABC_2d.py#L117

**Reply by shanmukhi1998** [reply]
I have been keeping it as [1,1] so far. Do you suggest any specific weights?

**Reply by pescap** [reply]
I would try `[1,weights]` with `weights = 10**i`, `i = 1,...,6`.

**Reply by shanmukhi1998** [reply]
Okay, thanks. I will try these weights once and get back to you.

**Reply by shanmukhi1998** [reply]
Hi @pescap and @lululxvi . I tried using different loss weights. I ran the model for 10,000 iterations. Just to reiterate, I am training a [3]-[18]*16-[1] model. I have around 500,000 data points. This time, I have included num_domain = 10000 and num_test = 20000 and also anchors = full data set. This is a slightly different data set and the PDE is the 2D transient heat diffusion equation (without convection and heat generation). I am training the data points that do not lie on the boundary (so as to not account for boundary effects). 
[LW-1e2.txt](https://github.com/lululxvi/deepxde/files/11502275/LW-1e2.txt)
[LW-1e6.txt](https://github.com/lululxvi/deepxde/files/11502276/LW-1e6.txt)
[LW-1e5.txt](https://github.com/lululxvi/deepxde/files/11502277/LW-1e5.txt)
[LW-1e4.txt](https://github.com/lululxvi/deepxde/files/11502278/LW-1e4.txt)
[LW-1e3.txt](https://github.com/lululxvi/deepxde/files/11502279/LW-1e3.txt)
![LW-1e2](https://github.com/lululxvi/deepxde/assets/122904243/545ba2e5-821c-4478-b95c-d9fbc3f514e3)
![LW-1e3](https://github.com/lululxvi/deepxde/assets/122904243/40d0c256-a423-49d1-98d8-965dc78530c3)
![LW-1e4](https://github.com/lululxvi/deepxde/assets/122904243/278d369f-99cd-4ffd-96e1-617ec571a823)
![LW-1e5](https://github.com/lululxvi/deepxde/assets/122904243/29c5a433-7c8f-4ea6-9772-caeb94e56abc)


I need your guidance here. How do I judge which one is doing better? Also, is it a good idea for PDE loss to not decrease continuously? Also, the MSE loss does not change after some time. Do you think, it is changing but in very, very small quantities? How do I get the convergence of the order of 1e-4? I will really appreciate your help here. I apologize for asking too many questions. Thanking you in advance.

**Reply by shanmukhi1998** [reply]
Apologies for any confusion. First image corresponds to 1e2, the second one to 1e3, the third one to 1e4 and the last one to 1e5. Also, for your kind reference, I have attached my code here:
[Code.txt](https://github.com/lululxvi/deepxde/files/11502341/Code.txt)
